milli/src/search/criteria/geo.rs

```diff
@@ -0,0 +1,117 @@
+    fn next(&mut self, params: &mut CriterionParameters) -> Result<Option<CriterionResult>> {
+        let rtree = self.rtree.as_ref();
+
+        loop {
+            match self.candidates.next() {
+                Some(mut candidates) => {
+                    candidates -= params.excluded_candidates;
+                    self.allowed_candidates -= &candidates;
+                    return Ok(Some(CriterionResult {
+                        query_tree: None,
+                        candidates: Some(candidates),
+                        filtered_candidates: None,
+                        bucket_candidates: Some(self.bucket_candidates.clone()),
+                    }));
+                }
+                None => match self.parent.next(params)? {
+                    Some(CriterionResult {
+                        query_tree,
+                        candidates,
+                        filtered_candidates,
+                        bucket_candidates,
+                    }) => {
+                        let mut candidates = match (&query_tree, candidates) {
+                            (_, Some(candidates)) => candidates,
+                            (Some(qt), None) => {
+                                let context = CriteriaBuilder::new(&self.rtxn, &self.index)?;
+                                resolve_query_tree(&context, qt, params.wdcache)?
+                            }
+                            (None, None) => self.index.documents_ids(self.rtxn)?,
+                        };
+
+                        if let Some(filtered_candidates) = filtered_candidates {
+                            candidates &= filtered_candidates;
+                        }
+
+                        match bucket_candidates {
+                            Some(bucket_candidates) => self.bucket_candidates |= bucket_candidates,
+                            None => self.bucket_candidates |= &candidates,
+                        }
+
+                        if candidates.is_empty() {
+                            continue;
+                        }
+                        self.allowed_candidates = &candidates - params.excluded_candidates;
+                        self.candidates = match rtree {
+                            Some(rtree) => {
+                                geo_point(rtree, self.allowed_candidates.clone(), self.point)
+                            }
+                            None => Box::new(std::iter::empty()),
+                        };
+                    }
+                    None => return Ok(None),
+                },
+            }
+        }
+    }
+}
+
+fn geo_point(
+    rtree: &RTree<GeoPoint>,
+    candidates: RoaringBitmap,
+    point: [f64; 2],
+) -> Box<dyn Iterator<Item = RoaringBitmap>> {
+    let results = rtree
+        .nearest_neighbor_iter(&point)
+        .filter_map(move |point| candidates.contains(point.data).then(|| point.data))
+        .map(|id| iter::once(id).collect::<RoaringBitmap>())
+        .collect::<Vec<_>>();
```

Comment:
We should stop iterating through the whole list of points, instead, we should stop when all the candidates have been found. By doing a `candidates.remove().is_some()` instead of a simple contains.

Reply:
Done thank you!